xline/src/server/lease_server.rs

```diff
@@ -123,13 +135,28 @@ where
     }
 
     /// Handle keep alive at leader
+    #[allow(clippy::integer_arithmetic)] // Introduced by tokio::select!
     async fn leader_keep_alive(
         &self,
         mut request_stream: tonic::Streaming<LeaseKeepAliveRequest>,
+        mut shutdown_listener: shutdown::Listener,
     ) -> Pin<Box<dyn Stream<Item = Result<LeaseKeepAliveResponse, tonic::Status>> + Send>> {
         let lease_storage = Arc::clone(&self.lease_storage);
         let stream = try_stream! {
-            while let Some(keep_alive_req) = request_stream.message().await? {
+           loop {
+                let keep_alive_req: LeaseKeepAliveRequest = tokio::select! {
+                    _ = shutdown_listener.wait_self_shutdown() => {
+                        debug!("Lease keep alive shutdown");
+                        break;
+                    }
+                    res = request_stream.message() => {
+                        if let Ok(Some(keep_alive_req)) = res {
+                            keep_alive_req
+                        } else {
+                            break;
+                        }
+                    }
+                };
                 debug!("Receive LeaseKeepAliveRequest {:?}", keep_alive_req);
                 let ttl = if lease_storage.is_primary() {
                     lease_storage.wait_synced(keep_alive_req.id).await;
```

Comment:
Will this wait_syned eventually complete?

Reply:
This may never be done when shutting down, Fixed